rasterio/_warp.pyx

```diff
@@ -652,56 +659,76 @@ def _calculate_default_transform(src_crs, dst_crs, width, height,
     elif isinstance(src_crs, dict):
         src_crs = CRS(**src_crs)
 
-    vrt_doc = _suggested_proxy_vrt_doc(width, height, transform=transform, crs=src_crs, gcps=gcps).decode('ascii')
+    # The transformer at the heart of this function requires a dataset.
+    # We use an in-memory VRT dataset.
+    vrt_doc = _suggested_proxy_vrt_doc(
+        width,
+        height,
+        transform=transform,
+        crs=src_crs,
+        gcps=gcps
+    ).decode('ascii')
+    hds = open_dataset(vrt_doc, 0x00 | 0x02 | 0x04, ['VRT'], {}, None)
```

Comment:
vrt_doc and hds assignments are better read when they are adjacent.

Reply:
this means the `open_dataset` is outside the `try expect` block, could this lead to issues?

Reply:
It shouldn't. open_dataset doesn't have any allocated C arguments, so there's nothing to leak if it fails.